subsys/bluetooth/controller/ll_sw/openisa/hal/RV32M1/radio/radio.c

```diff
@@ -955,28 +983,248 @@ u32_t radio_tmr_sample_get(void)
 	return 0;
 }
 
+void *radio_ccm_rx_pkt_set_ut(struct ccm *ccm, u8_t phy, void *pkt)
+{
+	/* Saved by LL as MSO to LSO in the ccm->key
+	 * SK (LSO to MSO)
+	 * :0x66:0xC6:0xC2:0x27:0x8E:0x3B:0x8E:0x05
+	 * :0x3E:0x7E:0xA3:0x26:0x52:0x1B:0xAD:0x99
+	 */
+	u8_t key_local[16] __aligned(4) = {
+		0x99, 0xad, 0x1b, 0x52, 0x26, 0xa3, 0x7e, 0x3e,
+		0x05, 0x8e, 0x3b, 0x8e, 0x27, 0xc2, 0xc6, 0x66
+	};
+	void *result;
+
+	/* ccm.key[16] is stored in MSO format, as retrieved from e function */
+	memcpy(ccm->key, key_local, sizeof(key_local));
+
+	/* Input std sample data, vol 6, part C, ch 1 */
+	_pkt_scratch[0] = 0x0f;
+	_pkt_scratch[1] = 0x05;
+	_pkt_scratch[2] = 0x9f; /* cleartext = 0x06*/
+	_pkt_scratch[3] = 0xcd;
+	_pkt_scratch[4] = 0xa7;
+	_pkt_scratch[5] = 0xf4;
+	_pkt_scratch[6] = 0x48;
+
+	/* IV std sample data, vol 6, part C, ch 1, stored in LL in LSO format
+	 * IV (LSO to MSO)      :0x24:0xAB:0xDC:0xBA:0xBE:0xBA:0xAF:0xDE
+	 */
+	ccm->iv[0] = 0x24;
+	ccm->iv[1] = 0xAB;
+	ccm->iv[2] = 0xDC;
+	ccm->iv[3] = 0xBA;
+	ccm->iv[4] = 0xBE;
+	ccm->iv[5] = 0xBA;
+	ccm->iv[6] = 0xAF;
+	ccm->iv[7] = 0xDE;
+
+	result = radio_ccm_rx_pkt_set(ccm, phy, pkt);
+	radio_ccm_is_done();
+
+	if (ctx_ccm.auth_mic_valid == 1 && ((u8_t *)pkt)[2] == 0x06) {
+		BT_INFO("Passed decrypt\n");
+	} else {
+		BT_INFO("Failed decrypt\n");
+	}
+
+	return result;
+}
+
 void *radio_ccm_rx_pkt_set(struct ccm *ccm, u8_t phy, void *pkt)
 {
-	printk("%s\n", __func__);
-	return NULL;
+	u8_t key_local[16] __aligned(4);
+	status_t status;
+	cau3_handle_t handle = {
+			.keySlot = kCAU3_KeySlot2,
+			.taskDone = kCAU3_TaskDonePoll
+	};
+	ARG_UNUSED(phy);
+
+	/* ccm.key[16] is stored in MSO format, as retrieved from e function */
+	memcpy(key_local, ccm->key, sizeof(key_local));
+	ctx_ccm.auth_mic_valid = 0;
+	ctx_ccm.empty_pdu_rxed = 0;
+	ctx_ccm.rx_pkt_in = (struct pdu_data *)_pkt_scratch;
+	ctx_ccm.rx_pkt_out = (struct pdu_data *)pkt;
+	ctx_ccm.nonce[0] = ccm->counter;	/* LSO to MSO, counter is LE */
```

Comment:
The type of the nonce is a bit wierd. It is being used as an array of 2 u64_t as well as a u8_t *, would it make more sense to have it as a byte array?
nonce[1] appears to be partially used. 
memcpy(nonce+5, , 8) would mean top bytes are never written). Is this ok?

Reply:
Initially, I was more interested in preserving the natural alignment of the counter, at 64 bits and be able to write that section of nonce with just one transfer without further penalties. By defining nonce as a group of 64-bits, I also made the ctx_ccm to have the proper alignment. If defined as bytearray, I would have needed an extra align in its definition.
The drawback is that the second 64-bit element needs a few extra casts to write into it. I will see if it can get cleaner.

As for your question about the last 3 bytes, thanks for taking the time. No, it is not a problem as, unfortunately, the driver for that peripheral rebuilds the B0 by copying nonce into the B0 block format. So, it does not look into the last 3 bytes for the storage in which nonce is passed. Some optimisations are planned in this area, but that is it for now.
Here is the action much lower in the driver. Nonce is the parameter from LLL/radio.c, L is computed to be 2, and copies nonce into b[1:13], and b[14:15] = sz.

    b[0] = authInSz ? 0x40 : 0 /* Adata */;
    b[0] |= (((authTagSz - 2) / 2) /* M' */<< 3);
    b[0] |= (L - 1) /* L' */;
    cau3_memcpy(&b[1], nonce, 15 - L);

    for (i = 0; i < L; i++)
    {
        if (mask && i >= wordSz)
            mask = 0x00;
        b[CAU3_AES_BLOCK_SIZE - 1 - i] = (uint8_t)((inSz >> ((8 * i) & mask))
                & mask);
    }

Reply:
I plan to use this construct for nonce. Let me know how does it feel like. I will update the PR.
static struct {
	union {
		u64_t counter;
		u8_t bytes[CAU3_AES_BLOCK_SIZE - 1 - 2];
	} nonce;	/* used by the B0 format but not in-situ */
	struct pdu_data *rx_pkt_out;
	struct pdu_data *rx_pkt_in;
	u8_t auth_mic_valid;
	u8_t empty_pdu_rxed;
} ctx_ccm;

Reply:
That would be more clear. I think it is a good change.